### PR TITLE
Use ExecutorService and use GitPullRequestRefsChanged

### LINKED ISSUE #54: Stop publishing webhooks on event threads
I'm the principal developer for Bitbucket Server.

Bitbucket Server has a constrained number of threads which handle invoking `@EventListener` methods (by default 1 per CPU core), and limits the pending event queue. Event listeners should not perform expensive operations directly; they should offload such processing to an `ExecutorService` or other mechanism so they can release the event thread as quickly as possible.

In addition to actually performing HTTP requests directly on the event thread, `PullRequestListener.rescopedEvent` is calling `PullRequestService.canMerge` on the event thread. If the pull request's refs are not up-to-date on disk, this may trigger a full `git merge` to update them (The `rescopedEvent` handler is explicitly seeking that behavior, actually), and before that can be done a _cluster-wide_ lock must be acquired. For Data Center customers, this is _extremely likely_ to trigger event thread starvation and can significantly impact many other operations.

This plugin can either create and manage its own `ExecutorService`, to allow it to control concurrency, or it can simply add an `@ComponentImport ExecutorService executorService` constructor parameter on listeners and use the one the host application exports for plugin developers. (That `ExecutorService` is also exported for the `Executor` and `ScheduledExecutorService` interfaces; you can `@ComponentImport` whichever you prefer.) The `ExecutorService` provided by the host application automatically propagates some state from the submitting thread to the executor thread that processes the operation (for example, the authenticated user and request details). To get the same special handling for your own `ExecutorService` you'll need to `@ComponentImport` SAL's `ThreadLocalDelegateExecutorFactory` and use the appropriate factory method to wrap your `ExecutorService`.

Please let me know if you have any questions, or if there's anything I can do to facilitate!

### LINKED ISSUE #56: PullRequestListener.rescopedEvent calls canMerge for declined pull requests
`PullRequestListener.rescopedEvent` calls `PullRequestService.canMerge` without first checking the state of the pull request on the event. When a pull request is rescoped and declined "close" together, it triggers errors like this:

```
2017-08-02 09:02:55,847 ERROR [AtlassianEvent::thread-9]  c.a.s.i.e.AsyncBatchingInvokersTransformer There was an exception thrown trying to dispatch event 'com.atlassian.bitbucket.event.pull.PullRequestRescopedEvent[source=com.atlassian.stash.internal.pull.DefaultPullRequestService@2a0b98d]' for the invoker 'SingleParameterMethodListenerInvoker{method=public void nl.topicus.bitbucket.api.PullRequestListener.rescopedEvent(com.atlassian.bitbucket.event.pull.PullRequestRescopedEvent) throws java.io.IOException, listener=nl.topicus.bitbucket.api.PullRequestListener@2fc0a10}'
java.lang.RuntimeException: This pull request has been declined and must be reopened before it can be merged.. Listener: nl.topicus.bitbucket.api.PullRequestListener event: com.atlassian.bitbucket.event.pull.PullRequestRescopedEvent
	at com.atlassian.event.internal.SingleParameterMethodListenerInvoker.invoke(SingleParameterMethodListenerInvoker.java:55)
	at com.atlassian.stash.internal.event.AsyncBatchingInvokersTransformer$AsyncInvokerBatch.invoke(AsyncBatchingInvokersTransformer.java:109)
	at com.atlassian.event.internal.AsynchronousAbleEventDispatcher$1$1.run(AsynchronousAbleEventDispatcher.java:38)
	at com.atlassian.sal.core.executor.ThreadLocalDelegateRunnable.run(ThreadLocalDelegateRunnable.java:34)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.lang.Thread.run(Thread.java:748)
	... 1 frame trimmed
Caused by: com.atlassian.bitbucket.pull.IllegalPullRequestStateException: This pull request has been declined and must be reopened before it can be merged.
	at com.atlassian.stash.internal.pull.DefaultPullRequestService.checkIsNotDeclined(DefaultPullRequestService.java:1223)
	at com.atlassian.stash.internal.pull.DefaultPullRequestService.checkIsOpen(DefaultPullRequestService.java:1235)
	at com.atlassian.stash.internal.pull.DefaultPullRequestService.canMerge(DefaultPullRequestService.java:244)
	at com.atlassian.plugin.util.ContextClassLoaderSettingInvocationHandler.invoke(ContextClassLoaderSettingInvocationHandler.java:26)
	at org.eclipse.gemini.blueprint.service.importer.support.internal.aop.ServiceInvoker.doInvoke(ServiceInvoker.java:56)
	at org.eclipse.gemini.blueprint.service.importer.support.internal.aop.ServiceInvoker.invoke(ServiceInvoker.java:60)
	at org.eclipse.gemini.blueprint.service.util.internal.aop.ServiceTCCLInterceptor.invokeUnprivileged(ServiceTCCLInterceptor.java:70)
	at org.eclipse.gemini.blueprint.service.util.internal.aop.ServiceTCCLInterceptor.invoke(ServiceTCCLInterceptor.java:53)
	at org.eclipse.gemini.blueprint.service.importer.support.LocalBundleContextAdvice.invoke(LocalBundleContextAdvice.java:57)
	at nl.topicus.bitbucket.api.PullRequestListener.rescopedEvent(PullRequestListener.java:94)
	at com.atlassian.event.internal.SingleParameterMethodListenerInvoker.invoke(SingleParameterMethodListenerInvoker.java:40)
	... 6 common frames omitted
```

Before calling `canMerge`, `rescopedEvent` should first check `isOpen()` on the event's `PullRequest`. If that's `true`, then it can call `canMerge`. If not, then it shouldn't. It should _also_ explicitly handle `IllegalPullRequestStateException` because, even with the check, there's always going to be room for race conditions between pull request updates and event processing.

### PR DESCRIPTION
**Pull request by casz**
Initial thoughts, any input?

Perhaps we should consider new event types for the onGitPullRequestRefsChanged.

To preserve other users of this webhook plugin just as an example:
https://github.com/bors-ng/bors-ng/issues/140

Fixes #53
Fixes #54  
Fixes #56

### CODE REVIEW

**Comment by casz** [comment]
@stephenc @bturner apparently I cannot request reviews.

File: src/main/java/nl/topicus/bitbucket/api/PullRequestListener.java [hunk 1]
```diff
@@ -127,43 +160,73 @@ private void sendPullRequestEvent(PullRequestEvent event, EventType eventType) t
...
+        });
+        submit(task);
+    }
+
+    private void submit(Callable<Void> task)
+    {
+        Future future = executorService.submit(task);
+        try
+        {
+            if (future.isDone()) {
```
Comment (abishek3876) [inline src/main/java/nl/topicus/bitbucket/api/PullRequestListener.java]: Why is this required?

File: src/main/java/nl/topicus/bitbucket/api/PullRequestListener.java [hunk 2]
```diff
@@ -127,43 +160,73 @@ private void sendPullRequestEvent(PullRequestEvent event, EventType eventType) t
...
-                        jsonBody,
-                        e);
+
             }
+            return null;
+        });
+        submit(task);
+    }
 
+    private void canMerge(PullRequest pullRequest)
```
Comment (abishek3876) [inline src/main/java/nl/topicus/bitbucket/api/PullRequestListener.java]: This may not reliably complete before we send out the event right?

File: src/main/java/nl/topicus/bitbucket/api/PullRequestListener.java (hunk 1 above)
Comment (casz) [inline src/main/java/nl/topicus/bitbucket/api/PullRequestListener.java]: Catching exception and logging them, I hadn't bothered with properly logging just yet.

It should properly be a while loop with some Thread.sleep to avoid blocking the main thread.

File: src/main/java/nl/topicus/bitbucket/api/PullRequestListener.java (hunk 2 above)
Comment (casz) [inline src/main/java/nl/topicus/bitbucket/api/PullRequestListener.java]: True, but that's where the new onGitPullRequestRefsChanged which is what Jenkins should be listening for which is why I suggest we create new EventTypes so automation servers can listen for these new events when bitbucket tells them it is ready.

And other dependants might still be interested in PR created events.

I would suggest a `EventType.GIT_PULL_REQUEST_UPDATED` and `EventType.GIT_PULL_REQUEST_CREATED`

Does that make sense?

File: src/main/java/nl/topicus/bitbucket/api/PullRequestListener.java (hunk 1 above)
Comment (casz) [inline src/main/java/nl/topicus/bitbucket/api/PullRequestListener.java]: Thinking about it, maybe not.

**Comment by casz** [comment]
Just so it doesn't go lost.

abishek3876: This may not reliably complete before we send out the event right?

me:
True, but that's where the new onGitPullRequestRefsChangedEvent comes in.
Which is what Jenkins should be listening for.
Hence why I suggest we create new EventTypes, so automation servers can listen for these new events when bitbucket tells them it is ready.

And other dependants might still be interested in PR created events and rescoped events without the canMerge being done such as Slack, Teams, Discord, Bots.

I would suggest an `EventType.GIT_PULL_REQUEST_UPDATED` and `EventType.GIT_PULL_REQUEST_CREATED`

Does that make sense?

**Comment by Eernie** [comment]
So because we check the `canMerge` we don't have to send webhooks anymore for the `rescoped` event? 
The `canMerge` will trigger a `GitPullRequestRefsChangedEvent` right?

**Comment by bturner** [comment]
Please don't merge. Typing a comment...

**Comment by bturner** [comment]
I reviewed yesterday, but didn't post my comments (intentionally), and I've reviewed the latest changes as well.

Considering the code, I think we're probably going the wrong way. Using the `GitPullRequestRefsChangedEvent` only makes sense if:
* We're willing to _not_ aggressively trigger merges, and/or
* We can infer the correct `EventType` from the state provided

The first is definitely not true. In order for the events to have the desired meaning, we _must_ trigger merges explicitly. If the second held true, though, it might still be worth doing.

Unfortunately, though, I don't think it does. Your fairly simple mapping between `RefChangeType` and `EventType` is not enough. For example:
* `ADD` can mean open _or reopen_, and there's no reliable way to tell the two apart short of searching the pull request's activity to check for a reopen event
* `DELETE` can mean declined _or merged_. Your code assumes declined where it really needs to check the `PullRequestState` explicitly

I think we're probably better off still calling `canMerge` (though ideally not the same way; more on that in a minute) and sending events after it completes, without looking at the Git-specific events.

I have a change made locally which shows I'd do this. I'm happy to push that up to a fork, if people would like to review it? (I've also fixed several other things while I was in the plugin's codebase.)

File: src/main/java/nl/topicus/bitbucket/api/PullRequestListener.java [hunk 3]
```diff
@@ -135,36 +163,50 @@ private void sendEvents(Object event, Repository repo, EventType eventType) thro
...
             }
+        };
+        executorService.execute(task);
+    }
 
-        }
+    private void canMerge(PullRequest pullRequest)
+    {
+        Runnable task = () -> {
+                if (pullRequest.isOpen()) {
```
Comment (bturner) [inline src/main/java/nl/topicus/bitbucket/api/PullRequestListener.java]: This is a step in the right direction, but it's not enough. You need something more like this:

```java
if (pullRequest.isOpen()) {
    try {
        pullRequestService.canMerge(pullRequest.getToRef().getRepository().getId(), pullRequest.getId());
    } catch (IllegalPullRequestStateException e) {
        LOGGER.debug("{}: {}", pullRequest, e.getMessage());
    }
}
```

(Note that passing `e.getMessage()` rather than `e` is _intentional_. We don't want to log the stack trace, because it's not useful.)
```

**Comment by casz** [comment]
Thanks for the commit. I see

Would love to review your changes.

**Comment by casz** [comment]
Rebased :)

File: src/main/java/nl/topicus/bitbucket/api/PullRequestListener.java [hunk 4]
```diff
@@ -135,36 +163,50 @@ private void sendEvents(Object event, Repository repo, EventType eventType) thro
...
+                    pullRequestService.canMerge(pullRequest.getToRef().getRepository().getId(), pullRequest.getId());
```
Comment (bturner) [inline src/main/java/nl/topicus/bitbucket/api/PullRequestListener.java]: Ideally we wouldn't call `PullRequestService.canMerge` to do this. That's also going to run `MergeRequestCheck`s, but we don't care about their _results_ (and based on support we know that there are instances which have _very expensive_ merge checks, which means throwing that data away "hurts" even more).

In a perfect world, we'd use [`ScmService.getPullRequestCommandFactory(PullRequest)`](https://developer.atlassian.com/static/javadoc/bitbucket-server/4.10.0/api/reference/com/atlassian/bitbucket/scm/ScmService.html#getPullRequestCommandFactory(com.atlassian.bitbucket.pull.PullRequest)) to get the `ScmPullRequestCommandFactory`, and we'd use that to trigger [`canMerge`](https://developer.atlassian.com/static/javadoc/bitbucket-server/4.10.0/api/reference/com/atlassian/bitbucket/scm/pull/ScmPullRequestCommandFactory.html#canMerge()) directly in the SCM.

You might have noticed I linked to 4.10.0 documentation, there. I did that because there's a wrinkle in trying to use `canMerge` directly. While the `PullRequestService.canMerge` API hasn't changed, the SCM's API _has_, and 4.10 was the release where it happened. In 4.9.x and earlier, you'd call `ScmPullRequestCommandFactory.canMerge`, but in 4.10+ you need to call [`tryMerge`](https://developer.atlassian.com/static/javadoc/bitbucket-server/4.10.0/api/reference/com/atlassian/bitbucket/scm/pull/ScmPullRequestCommandFactory.html#tryMerge(com.atlassian.bitbucket.pull.PullRequest)) instead.

For 4.x it doesn't matter. You could consistently call `canMerge` on every 4.x version. It matters for compatibility with 5.x, where `canMerge` has been completely removed and `tryMerge` is the only available method.

To do this directly in the SCM, we'll need to:
* Compile against 4.10
  * This does _not_ mean the plugin's compatibility range changes. It just means we get access to newer APIs.
* Check `ApplicationPropertiesService.getBuildVersion()` to determine the server version
  * This service has already been injected, so it's an easy check to add
* Call `canMerge` or `tryMerge` based on the detected version

I want to be clear that we _don't_ need to change this in _this pull request_. Technically we don't "need" to change it _at all_. But it will make the plugin perform significantly better on many instances if we _do_.

**Comment by bturner** [comment]
@casz Give me a moment to finish a couple tweaks and I'll give you a link to my commit. Thanks for taking the time to look it over!

**Comment by casz** [comment]
Some good comments 👍 

So go back to sending events on `rescopedEvent` and `createdEvent` but wait for the executor to tell us it is successfully done, is what I am getting.

Then in another PR look into calling `tryMerge`/`canMerge` 💯

**Comment by bturner** [comment]
@casz @Eernie I've raised #57 with these example changes. You're welcome to dissect it, merge it, ignore it as you see fit. If you have any questions, feel free to ask!

**Comment by casz** [comment]
Closing in favor of #57
